filewatcher/filewatcher.go

```diff
@@ -308,6 +305,26 @@ func New(ctx context.Context, cfg Config) (*Result, error) {
 	if err != nil {
 		return nil, err
 	}
+	isDir, err := isDirectory(filepath.Clean(cfg.Path))
+	if err != nil {
+		return nil, err
+	}
+	if isDir {
+		// Need to walk recursively because the watcher
+		// doesnt support recursion by itself
+		dirPath := filepath.Clean(cfg.Path)
+
+		err = filepath.WalkDir(dirPath, func(path string, info fs.DirEntry, err error) error {
+			if info.IsDir() {
+				return watcher.Add(path)
```

Comment:
I'm not sure that this works as you intended it to be.

so between filewatcher and directory watcher, the actual parser api would be different. for filewatcher since it's just a file, you can just pass the `io.Reader` to the parser to let it do the parsing. but for directory watcher you can't really get an `io.Reader` out of a directory, so what you want to do is just notify the caller that "file(s) under this directory changed, you want to reload everything", with potentially the path to the directory, or maybe `[]fs.DirEntry` (this is what's returned by `os.ReadDir`), passed into the parser callback.

also I think kyle's earlier suggestion of wait for an interval without any notifications before doing that is important, otherwise we could be calling the directory callback in very quick succession.

Reply:
So I need to check watch all of the underlying paths and act accordingly when events happen. And in this case the parser being used is `csiPathParser` so I do think it works similarly enough to the regular filewatcher.

How would you expect the timer to be implemented? I wanted to minimize changes to the behavior of the current watcherloop so I did it this way

Reply:
so if you look at the doc of [`time.AfterFunc`](https://pkg.go.dev/time#AfterFunc), it returns you a `*time.Timer` that you can `Reset` or `Stop`.

basically you want to store the `*time.Timer` returned by it in the `*Result` struct, along with a lock, so basically in `watcherLoop`, every time a file under a directory changes, you do something like this:

```go
if isDirectory(...) {
  // use a lambda to make sure `defer` works as expected. you can also just refactor this part into a function
  func() {
    r.dirTimerLock.Lock()
    defer r.dirTimerLock.Unlock()
    if r.dirTimer != nil {
      // this means this is not the first event in thischange, just reset the timer
      r.dirTimer.Reset(configuredDuration)
      return
    }
    // otherwise, this is the first event in this change and we need to create a new timer
    r.dirTimer = time.AfterFunc(configuredDuration, func() {
      // this is the function fired to call the parser and remove the stored timer
      r.dirTimerLock.Lock()
      r.dirTimerLock = nil
      r.dirTimerLock.Unlock()
      dirParser(...)
    })
  }()
}
```

basically, for every fsnotify event, we push the timer on `configuredDuration`, and if no new fsnotify event happens during this duration we actually call the parser to reload everything.